File: lib/services/vixrex_nlu/vixrex_conversation_context.dart
import 'package:vixrex/services/vixrex_nlu/vixrex_normalizer.dart';

class VixrexBekleyenBaglam {
  final String anahtar;
  final String etiket;
  final String tip;
  final String? eylem;

  const VixrexBekleyenBaglam({
    required this.anahtar,
    required this.etiket,
    required this.tip,
    this.eylem,
  });
}

enum VixrexBaglamKarari {
  genelSor,
  ozelSor,
  ayniKalsin,
  kaldirmaOnayi,
  kaldir,
  iptal,
  boolTrue,
  boolFalse,
  saatEksik,
  adresEksik,
  deger,
}

class VixrexBaglamSonucu {
  final VixrexBaglamKarari karar;
  final bool yazma;
  final String mesaj;
  final Object? deger;

  const VixrexBaglamSonucu({
    required this.karar,
    required this.yazma,
    required this.mesaj,
    this.deger,
  });
}

const vixrexDogalGenelSoru = 'Vitrininde neyi farklı görmek istersin?';

const _iptal = {'iptal', 'vazgec', 'vazgectim'};
const _evet = {'evet'};
const _hayir = {'hayir'};
const _ayni = {'aynisi', 'aynen aynisi'};
const _degerOlmayanKisa = {
  'tamam',
  'olur',
  'peki',
  'aynen',
  'degistir',
  'degissin',
  'yap',
};

String _alanSorusu(VixrexBekleyenBaglam bekleyen) =>
    '${bekleyen.etiket} için ne yazayım?';

bool _calismaSaatiTamMi(String input) {
  final norm = VixrexNormalizer.normalize(input);
  if (RegExp(r'\b24\s*saat\b').hasMatch(norm) ||
      RegExp(r'\b7\s*[/ ]\s*24\b').hasMatch(norm)) {
    return true;
  }

  final saatler = RegExp(r'\b(?:[01]?\d|2[0-3])[:.]\d{2}\b').allMatches(input);
  if (saatler.length >= 2) return true;

  if (RegExp(r'\b\d{1,2}\s*[-–—]\s*\d{1,2}\b').hasMatch(norm)) {
    return true;
  }

  final sayilar = RegExp(r'\b\d{1,2}\b').allMatches(norm).length;
  return sayilar >= 2 && norm.contains('sabah') && norm.contains('aksam');
}

bool _adresYeterinceAcikMi(String input) {
  final norm = VixrexNormalizer.normalize(input);
  final kelimeler = norm.split(RegExp(r'\s+')).where((e) => e.isNotEmpty).toList();
  if (kelimeler.length > 3) return true;
  return RegExp(
    r'\b(mahalle|mah|cadde|cad|sokak|sok|bulvar|blv|no|numara|apartman|apt|site|meydan)\b',
  ).hasMatch(norm);
}

/// Yalnız mevcut konuşma bağlamını yorumlar. Yeni niyet çözmez ve kayıt yapmaz.
/// Kısa cevap bağlamla güvenle açıklanamıyorsa vitrine yazmak yerine doğal bir
/// takip sorusu üretir.
VixrexBaglamSonucu vixrexBaglamsalCevapKarari(
  String input,
  VixrexBekleyenBaglam? bekleyen,
) {
  final trimmed = input.trim();
  final norm = VixrexNormalizer.normalize(trimmed);

  if (bekleyen == null) {
    return const VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.genelSor,
      yazma: false,
      mesaj: vixrexDogalGenelSoru,
    );
  }

  if (bekleyen.eylem == 'kaldir') {
    if (_evet.contains(norm)) {
      return const VixrexBaglamSonucu(
        karar: VixrexBaglamKarari.kaldir,
        yazma: true,
        mesaj: '',
        deger: '',
      );
    }
    if (_hayir.contains(norm) || _iptal.contains(norm)) {
      return VixrexBaglamSonucu(
        karar: VixrexBaglamKarari.iptal,
        yazma: false,
        mesaj: 'Tamam, ${bekleyen.etiket} bilgisini kaldırmıyorum.',
      );
    }
    return VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.kaldirmaOnayi,
      yazma: false,
      mesaj:
          '${bekleyen.etiket} bilgisini kaldırmamı istiyorsan evet, vazgeçtiysen hayır diyebilirsin.',
    );
  }

  if (_iptal.contains(norm)) {
    return const VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.iptal,
      yazma: false,
      mesaj: 'Tamam, bu değişikliği yapmıyorum. Başka neyi değiştirmek istersin?',
    );
  }

  if (_ayni.contains(norm)) {
    return VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.ayniKalsin,
      yazma: false,
      mesaj: 'Tamam, ${bekleyen.etiket} aynı kalsın.',
    );
  }

  if (bekleyen.tip == 'acikKapali') {
    if (_evet.contains(norm)) {
      return const VixrexBaglamSonucu(
        karar: VixrexBaglamKarari.boolTrue,
        yazma: true,
        mesaj: '',
        deger: true,
      );
    }
    if (_hayir.contains(norm)) {
      return const VixrexBaglamSonucu(
        karar: VixrexBaglamKarari.boolFalse,
        yazma: true,
        mesaj: '',
        deger: false,
      );
    }
  } else if (_evet.contains(norm) ||
      _hayir.contains(norm) ||
      _degerOlmayanKisa.contains(norm)) {
    return VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.ozelSor,
      yazma: false,
      mesaj: _alanSorusu(bekleyen),
    );
  }

  if (RegExp(r'\b(onu|bunu)\s+(kaldir|sil|temizle)\b').hasMatch(norm) ||
      RegExp(r'^(kaldir|sil|temizle)$').hasMatch(norm)) {
    return VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.kaldirmaOnayi,
      yazma: false,
      mesaj: '${bekleyen.etiket} bilgisini kaldırmamı mı istiyorsun?',
    );
  }

  if (bekleyen.anahtar == 'calismaSaatleri' && !_calismaSaatiTamMi(trimmed)) {
    return const VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.saatEksik,
      yazma: false,
      mesaj: 'Çalışma saatlerini tamamlamak için kaçta açıp kaçta kapandığınızı yazar mısın?',
    );
  }

  if (bekleyen.anahtar == 'adres' && !_adresYeterinceAcikMi(trimmed)) {
    return const VixrexBaglamSonucu(
      karar: VixrexBaglamKarari.adresEksik,
      yazma: false,
      mesaj: 'Açık adresi biraz daha ayrıntılı yazar mısın?',
    );
  }

  return VixrexBaglamSonucu(
    karar: VixrexBaglamKarari.deger,
    yazma: true,
    mesaj: '',
    deger: trimmed,
  );
}

File: public_web/src/lib/vixrexConversationContext.ts
import { vixrexNormalizeDartParity } from "./vixrexNormalizer";

export interface VixrexBekleyenBaglam {
  anahtar: string;
  etiket: string;
  tip: string;
  eylem?: "kaldir";
}

export type VixrexBaglamKarari =
  | "genel_sor"
  | "ozel_sor"
  | "ayni_kalsin"
  | "kaldirma_onayi"
  | "kaldir"
  | "iptal"
  | "bool_true"
  | "bool_false"
  | "saat_eksik"
  | "adres_eksik"
  | "deger";

export interface VixrexBaglamSonucu {
  karar: VixrexBaglamKarari;
  yazma: boolean;
  mesaj: string;
  deger?: string | boolean | null;
}

export const VIXREX_DOGAL_GENEL_SORU = "Vitrininde neyi farklı görmek istersin?";

const IPTAL = new Set(["iptal", "vazgec", "vazgectim"]);
const EVET = new Set(["evet"]);
const HAYIR = new Set(["hayir"]);
const AYNI = new Set(["aynisi", "aynen aynisi"]);
const DEGER_OLMAYAN_KISA = new Set([
  "tamam",
  "olur",
  "peki",
  "aynen",
  "degistir",
  "degissin",
  "yap",
]);

function alanSorusu(bekleyen: VixrexBekleyenBaglam): string {
  return `${bekleyen.etiket} için ne yazayım?`;
}

function calismaSaatiTamMi(input: string): boolean {
  const norm = vixrexNormalizeDartParity(input);
  if (/\b24\s*saat\b/.test(norm) || /\b7\s*[/ ]\s*24\b/.test(norm)) return true;

  const saatler = input.match(/\b(?:[01]?\d|2[0-3])[:.]\d{2}\b/g) ?? [];
  if (saatler.length >= 2) return true;

  if (/\b\d{1,2}\s*[-–—]\s*\d{1,2}\b/.test(norm)) return true;

  const sayilar = norm.match(/\b\d{1,2}\b/g) ?? [];
  if (sayilar.length >= 2 && norm.includes("sabah") && norm.includes("aksam")) return true;

  return false;
}

function adresYeterinceAcikMi(input: string): boolean {
  const norm = vixrexNormalizeDartParity(input);
  const kelimeler = norm.split(/\s+/).filter(Boolean);
  if (kelimeler.length > 3) return true;
  return /\b(mahalle|mah|cadde|cad|sokak|sok|bulvar|blv|no|numara|apartman|apt|site|meydan)\b/.test(norm);
}

/**
 * Yalnız mevcut konuşma bağlamını yorumlar. Yeni niyet çözmez ve kayıt yapmaz.
 * Amaç kısa cevapları önceki soruyla birlikte değerlendirmek; bağlam yetersizse
 * tahmin ederek vitrine yazmak yerine doğal bir takip sorusu üretmektir.
 */
export function vixrexBaglamsalCevapKarari(
  input: string,
  bekleyen: VixrexBekleyenBaglam | null,
): VixrexBaglamSonucu {
  const trimmed = input.trim();
  const norm = vixrexNormalizeDartParity(trimmed);

  if (!bekleyen) {
    return { karar: "genel_sor", yazma: false, mesaj: VIXREX_DOGAL_GENEL_SORU };
  }

  // Bir önceki turda kullanıcı açıkça “onu kaldır” dedi ve Asistan onay
  // sorduysa, sonraki evet/hayır artık alan değeri değildir; o kaldırma
  // kararının cevabıdır.
  if (bekleyen.eylem === "kaldir") {
    if (EVET.has(norm)) {
      return { karar: "kaldir", yazma: true, mesaj: "", deger: null };
    }
    if (HAYIR.has(norm) || IPTAL.has(norm)) {
      return {
        karar: "iptal",
        yazma: false,
        mesaj: `Tamam, ${bekleyen.etiket} bilgisini kaldırmıyorum.`,
      };
    }
    return {
      karar: "kaldirma_onayi",
      yazma: false,
      mesaj: `${bekleyen.etiket} bilgisini kaldırmamı istiyorsan evet, vazgeçtiysen hayır diyebilirsin.`,
    };
  }

  if (IPTAL.has(norm)) {
    return {
      karar: "iptal",
      yazma: false,
      mesaj: "Tamam, bu değişikliği yapmıyorum. Başka neyi değiştirmek istersin?",
    };
  }

  if (AYNI.has(norm)) {
    return {
      karar: "ayni_kalsin",
      yazma: false,
      mesaj: `Tamam, ${bekleyen.etiket} aynı kalsın.`,
    };
  }

  if (bekleyen.tip === "acikKapali") {
    if (EVET.has(norm)) {
      return { karar: "bool_true", yazma: true, mesaj: "", deger: true };
    }
    if (HAYIR.has(norm)) {
      return { karar: "bool_false", yazma: true, mesaj: "", deger: false };
    }
  } else if (EVET.has(norm) || HAYIR.has(norm) || DEGER_OLMAYAN_KISA.has(norm)) {
    return {
      karar: "ozel_sor",
      yazma: false,
      mesaj: alanSorusu(bekleyen),
    };
  }

  if (/\b(onu|bunu)\s+(kaldir|sil|temizle)\b/.test(norm) || /^(kaldir|sil|temizle)$/.test(norm)) {
    return {
      karar: "kaldirma_onayi",
      yazma: false,
      mesaj: `${bekleyen.etiket} bilgisini kaldırmamı mı istiyorsun?`,
    };
  }

  if (bekleyen.anahtar === "calismaSaatleri" && !calismaSaatiTamMi(trimmed)) {
    return {
      karar: "saat_eksik",
      yazma: false,
      mesaj: "Çalışma saatlerini tamamlamak için kaçta açıp kaçta kapandığınızı yazar mısın?",
    };
  }

  if (bekleyen.anahtar === "adres" && !adresYeterinceAcikMi(trimmed)) {
    return {
      karar: "adres_eksik",
      yazma: false,
      mesaj: "Açık adresi biraz daha ayrıntılı yazar mısın?",
    };
  }

  return { karar: "deger", yazma: true, mesaj: "", deger: trimmed };
}
